crux_time/src/lib.rs

```diff
@@ -162,23 +176,93 @@ where
 
     /// Ask to receive a notification when the specified duration has elapsed.
     /// This is an async call to use with [`crux_core::compose::Compose`].
-    pub async fn notify_after_async(&self, id: TimerId, duration: Duration) -> TimeResponse {
-        self.context
-            .request_from_shell(TimeRequest::NotifyAfter { id, duration })
-            .await
+    pub fn notify_after_async(
+        &self,
+        id: TimerId,
+        duration: Duration,
+    ) -> TimerFuture<impl Future<Output = TimeResponse>> {
+        let future = self
+            .context
+            .request_from_shell(TimeRequest::NotifyAfter { id, duration });
+        TimerFuture::new(id, future)
     }
 
     pub fn clear(&self, id: TimerId) {
         self.context.spawn({
-            let context = self.context.clone();
+            {
+                let mut lock = CLEARED_TIMER_IDS.lock().unwrap();
+                lock.insert(id);
+            }
 
+            let context = self.context.clone();
             async move {
                 context.notify_shell(TimeRequest::Clear { id }).await;
             }
         });
     }
 }
 
+pin_project_lite::pin_project! {
+    pub struct TimerFuture<F>
+    where
+        F: Future<Output = TimeResponse>,
+    {
+        timer_id: TimerId,
+        is_cleared: bool,
+        #[pin]
+        future: F,
+    }
+}
+
+impl<F> Future for TimerFuture<F>
+where
+    F: Future<Output = TimeResponse>,
+{
+    type Output = TimeResponse;
+
+    fn poll(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Self::Output> {
+        if self.is_cleared {
+            // short-circuit return
+            return Poll::Ready(TimeResponse::Cleared { id: self.timer_id });
+        };
+        // see if the timer has been cleared
+        let timer_is_cleared = {
+            let mut lock = CLEARED_TIMER_IDS.lock().unwrap();
+            lock.remove(&self.timer_id)
+        };
+        let this = self.project();
+        *this.is_cleared = timer_is_cleared;
```

Comment:
I'd have to try to have specific advice, but I think you might be able to avoid the `pin_project` by `self.deref_mut()` and an `Unpin` trait bound for whatever is complaining about not being guaranteed unpin. 

Was that problematic?

Reply:
Have removed `pin_project`, you're right, it wasn't necessary in this instance.